src/target/linux.rs

```diff
@@ -90,41 +94,88 @@ impl UnixTarget for LinuxTarget {
         self.pid
     }
 
-    fn unpause(&self) -> Result<nix::sys::wait::WaitStatus, Box<dyn std::error::Error>> {
+    fn unpause(&self) -> Result<WaitStatus, Box<dyn std::error::Error>> {
+        self.handle_breakpoint()?;
         ptrace::cont(self.pid(), None)?;
-        let status = nix::sys::wait::waitpid(self.pid(), None)?;
-        match status {
-            // We may have hit a user defined breakpoint
-            nix::sys::wait::WaitStatus::Stopped(_, nix::sys::signal::Signal::SIGTRAP) => {
-                if let Some(bp) = self
-                    .breakpoints
-                    .borrow_mut()
-                    .get_mut(&(self.read_regs()?.rip as usize - 1))
+        let status = waitpid(self.pid(), None)?;
+
+        //// We may have hit a user defined breakpoint
+        if let WaitStatus::Stopped(_, nix::sys::signal::Signal::SIGTRAP) = status {
+            if let Some(bp) = self
+                .breakpoints
+                .borrow_mut()
+                .get_mut(&(self.read_regs()?.rip as usize - 1))
+            {
+                // scope the borrow of hit_breakpoint
                 {
-                    // Restore the program to it's uninstrumented state
-                    self.restore_breakpoint(bp)?;
-                    // Find a way to let the user do things here
-                };
-            }
-            _ => (),
-        };
+                    // Register what breakpoint was  hit
+                    self.hit_breakpoint.borrow_mut().replace(bp.clone());
+                }
+
+                // Restore the program to it's uninstrumented state
+                self.restore_breakpoint(bp)?;
+            };
+        }
+        Ok(status)
+    }
+
+    fn step(&self) -> Result<WaitStatus, Box<dyn std::error::Error>> {
+        if !self.handle_breakpoint()? {
+            ptrace::step(self.pid(), None).map_err(|e| Box::new(e))?;
+        }
+        let status = waitpid(self.pid(), None)?;
+        assert_eq!(
+            status,
+            WaitStatus::Stopped(self.pid(), nix::sys::signal::SIGTRAP)
+        );
         Ok(status)
     }
 }
 
 impl LinuxTarget {
+    /// executes the code instrumented by a breakpoint if one has been hit
+    /// and resets the breakpoint
+    /// returns true if it executed and instruction, false otherwise
+    /// The function may not execute the instrumented instruction if the
+    /// P.C is not where the breakpoint was set
+    fn handle_breakpoint(&self) -> Result<bool, Box<dyn std::error::Error>> {
+        let mut stepped = false;
+
+        let bp = { self.hit_breakpoint.borrow().clone() };
```

Comment:
You could use `let bp = self.hit_breakpoint.borrow_mut().take()` here. That way you don't have to manually clear it.